File: platforms/macos/PwdlockMac/Sources/PwdlockCore/Security/VaultKeyEnvelope.swift
import CryptoKit
import Foundation

public enum VaultKeyEnvelopeError: Error, Equatable {
    case invalidVaultKey
    case authenticationFailed
}

public enum VaultKeyEnvelope {
    public static func wrap(
        vaultKey: Data,
        masterPassword: String,
        vaultID: Data,
        passwordSalt: Data,
        wrapNonce: Data,
        parameters: Argon2idParameters
    ) throws -> VaultMetadata {
        guard vaultKey.count == 32 else {
            throw VaultKeyEnvelopeError.invalidVaultKey
        }

        let template = VaultMetadata(
            vaultID: vaultID,
            memoryKiB: parameters.memoryKiB,
            iterations: parameters.iterations,
            parallelism: UInt8(exactly: parameters.parallelism) ?? 0,
            passwordSalt: passwordSalt,
            wrapNonce: wrapNonce,
            wrappedVaultKey: Data(repeating: 0, count: 32),
            wrapTag: Data(repeating: 0, count: 16)
        )
        let kek = try Argon2id.deriveKey(
            password: PasswordInput.utf8NFC(masterPassword),
            salt: passwordSalt,
            parameters: parameters
        )
        let nonce = try AES.GCM.Nonce(data: wrapNonce)
        let sealed = try AES.GCM.seal(
            vaultKey,
            using: SymmetricKey(data: kek),
            nonce: nonce,
            authenticating: VaultMetadataCodec.authenticatedHeader(for: template)
        )

        return VaultMetadata(
            vaultID: vaultID,
            memoryKiB: parameters.memoryKiB,
            iterations: parameters.iterations,
            parallelism: UInt8(exactly: parameters.parallelism) ?? 0,
            passwordSalt: passwordSalt,
            wrapNonce: wrapNonce,
            wrappedVaultKey: sealed.ciphertext,
            wrapTag: sealed.tag
        )
    }

    public static func unwrap(_ metadata: VaultMetadata, masterPassword: String) throws -> Data {
        do {
            let parameters = Argon2idParameters(
                memoryKiB: metadata.memoryKiB,
                iterations: metadata.iterations,
                parallelism: UInt32(metadata.parallelism)
            )
            let kek = try Argon2id.deriveKey(
                password: PasswordInput.utf8NFC(masterPassword),
                salt: metadata.passwordSalt,
                parameters: parameters
            )
            let box = try AES.GCM.SealedBox(
                nonce: AES.GCM.Nonce(data: metadata.wrapNonce),
                ciphertext: metadata.wrappedVaultKey,
                tag: metadata.wrapTag
            )
            return try AES.GCM.open(
                box,
                using: SymmetricKey(data: kek),
                authenticating: VaultMetadataCodec.authenticatedHeader(for: metadata)
            )
        } catch {
            throw VaultKeyEnvelopeError.authenticationFailed
        }
    }
}

File: platforms/android/app/src/main/java/com/pwdlock/android/crypto/VaultKeyEnvelope.kt
package com.pwdlock.android.crypto

class VaultKeyEnvelopeException(message: String, cause: Throwable? = null) : Exception(message, cause)

/**
 * 复刻 macOS `VaultKeyEnvelope`：用主密码把 32 字节 vaultKey 包装 / 解包成 [VaultMetadata]。
 *
 * - KEK = Argon2id(主密码, passwordSalt, 参数)
 * - AES-GCM 以 wrapNonce 加密 vaultKey，AAD 为 metadata 编码后的前 52 字节
 * - 密码错误或数据损坏统一抛 [VaultKeyEnvelopeException]
 */
object VaultKeyEnvelope {
    fun wrap(
        vaultKey: ByteArray,
        masterPassword: String,
        vaultID: ByteArray,
        passwordSalt: ByteArray,
        wrapNonce: ByteArray,
        params: Argon2id.Params = Argon2id.initial,
    ): VaultMetadata {
        require(vaultKey.size == 32) { "vaultKey must be 32 bytes" }
        val template = VaultMetadata(
            vaultID = vaultID,
            memoryKiB = params.memoryKiB,
            iterations = params.iterations,
            parallelism = params.parallelism,
            passwordSalt = passwordSalt,
            wrapNonce = wrapNonce,
            wrappedVaultKey = ByteArray(32),
            wrapTag = ByteArray(16),
        )
        val kek = Argon2id.deriveKey(masterPassword, passwordSalt, params)
        val sealed = AesGcm.seal(vaultKey, kek, wrapNonce, VaultMetadataCodec.authenticatedHeader(template))
        val (ct, tag) = AesGcm.split(sealed)
        return template.copy(wrappedVaultKey = ct, wrapTag = tag)
    }

    fun unwrap(metadata: VaultMetadata, masterPassword: String): ByteArray = try {
        val params = Argon2id.Params(metadata.memoryKiB, metadata.iterations, metadata.parallelism)
        val kek = Argon2id.deriveKey(masterPassword, metadata.passwordSalt, params)
        val sealed = AesGcm.join(metadata.wrappedVaultKey, metadata.wrapTag)
        AesGcm.open(sealed, kek, metadata.wrapNonce, VaultMetadataCodec.authenticatedHeader(metadata))
    } catch (e: Exception) {
        throw VaultKeyEnvelopeException("authentication failed", e)
    }
}
